app/screens/login/form.tsx

```diff
@@ -0,0 +1,394 @@
+
+    const focus = () => {
+        if (Platform.OS === 'ios') {
+            let ssoOffset = 0;
+            switch (numberSSOs) {
+                case 0:
+                    ssoOffset = 0;
+                    break;
+                case 1:
+                case 2:
+                    ssoOffset = 48;
+                    break;
+                default:
+                    ssoOffset = 3 * 48;
+                    break;
+            }
+            let offsetY = 150 - ssoOffset;
+            if (isTablet) {
+                const {width, height} = dimensions;
+                const isLandscape = width > height;
+                offsetY = (isLandscape ? 230 : 150) - ssoOffset;
+            }
+            requestAnimationFrame(() => {
+                keyboardAwareRef.current?.scrollToPosition(0, offsetY);
+            });
+        }
+    };
+
+    const preSignIn = preventDoubleTap(async () => {
+        setIsLoading(true);
+
+        Keyboard.dismiss();
+        signIn();
+    });
+
+    const signIn = async () => {
+        const result: LoginActionResponse = await login(serverUrl!, {serverDisplayName, loginId: loginId.toLowerCase(), password, config, license});
+        if (checkLoginResponse(result)) {
+            if (!result.hasTeams && !result.error) {
+                // eslint-disable-next-line no-console
+                console.log('TODO: GO TO NO TEAMS');
+                return;
+            }
+            goToHome(result.time || 0, result.error as never);
+        }
+    };
+
+    const goToHome = (time: number, loginError?: never) => {
+        const hasError = launchError || Boolean(loginError);
+        resetToHome({extra, launchError: hasError, launchType, serverUrl, time});
+    };
+
+    const checkLoginResponse = (data: LoginActionResponse) => {
+        let errorId = '';
+        const clientError = data.error as ClientErrorProps;
+        if (clientError && clientError.server_error_id) {
+            errorId = clientError.server_error_id;
+        }
+
+        if (data.failed && MFA_EXPECTED_ERRORS.includes(errorId)) {
+            goToMfa();
+            setIsLoading(false);
+            return false;
+        }
+
+        if (data?.error && data.failed) {
+            setIsLoading(false);
+            setError(getLoginErrorMessage(data.error));
+            return false;
+        }
+
+        setIsLoading(false);
+
+        return true;
+    };
+
+    const goToMfa = () => {
+        goToScreen(MFA, '', {goToHome, loginId, password, config, serverDisplayName, license, serverUrl, theme}, loginAnimationOptions());
+    };
+
+    const getLoginErrorMessage = (loginError: string | ClientErrorProps | Error) => {
+        if (typeof loginError === 'string') {
+            return loginError;
+        }
+
+        const serverError = loginError as ClientErrorProps;
+        const errorId = serverError.server_error_id;
```

Comment:
If `loginError` is of type `Error`, then we won't have the `server-error_id` property.  
I think that this function can be flattened but 0/5 .
```
        const errorId = serverError.server_error_id;
```

Reply:
what would be your suggestion? your worry is being addressed on line 180